### ISSUE TITLE
No generic type on SourceFileBase in typings

### ISSUE THREAD

**Issue by LeonardKoch** [dsherret/ts-morph#624]
**Describe the bug**

Version: 2.1.0

Excerpt from the typings starting at line 7475:
```
declare const SourceFileBase: Constructor<ModuledNode> & Constructor<StatementedNode> & Constructor<TextInsertableNode> & typeof Node;

export declare class SourceFile extends SourceFileBase<ts.SourceFile> {
```

When SourceFile is declared it is said to extend SourceFileBase of generic type ts.SourceFile.
But accordinmg to the declaration above SourceFileBase does not support a generic type.
This leads to to for example Intellij/Webstorm to not include any of the properties from SourceFileBase in its completion for SourceFile.
Many other tools do support this and so this might also be an issue on JetBrains side, but I thought it worth bringing up to make sure this was intentional.

**To Reproduce**

Adding steps to reproduce doesn't make 100% sense here but I'm still going to do it.

Add the code below into a TSMorph project and open it in IntelliJ or Webstorm.
```ts
import { Project } from "ts-morph";

const project = new Project();
const sourceFile = project.createSourceFile("test.ts", ``);
sourceFile.addVariableStatement({
        declarationKind: VariableDeclarationKind.Const,
        declarations: [{
            name: "myNumber",
            initializer: "5"
        }
    });
```
The IDE will not understand that addVariableStatement is present on SourceFile.
If you now go into the typings for TS Morph and on line 7477 remove the generic type <ts.SourceFile>, it will recognize that addVariableStatement and various other properties are present on SourceFile.

**Expected behavior**

Expected behavior is that the methods from ModuledNode, StatementedNode & TextInsertableNode show up in completion in all IDEs, but it might also be that IntelliJ has a bug here ;)
Sorry if this issue is nuisance and `SourceFileBase<ts.SourceFile>` makes total sense.
Thanks a lot for building this great library.

**Comment by dsherret** [comment]
Interesting. `SourceFileBase` is a variable declaration so it can't have a type parameter. Seems like a bug in intelliji/webstorm as this works fine with typescript's language service (I think jetbrains does their own thing and they probably don't handle this scenario).

That said, perhaps for the sake of the typings I could transform the code when emitting the declaration file so this works. Not sure what I could do, but I'll look into it.

On a sidenote, I used to use resharper (also jetbrains) a couple years ago with TypeScript and I had to drop it because there were so many bugs and it was super slow. Then eventually I had to drop visual studio entirely because it was slow and use visual studio code. How has your experience been with Intellij/Webstorm and typescript other than this bug?

**Comment by LeonardKoch** [comment]
Hey,
thanks for responding to this so quickly David.
I'll definitely raise an issue with Jetbrains about fixing this in their TS implementation.
They do this funky thing where they have their own service for completions, but use the TS Language Service for error reporting.

This is really quite interesting, because the code makes total sense in the source:
From the ts-morph source:
src/compiler/ast/module/SourceFile.ts Line 39-43
```
// todo: not sure why I need to explicitly type this in order to get TS to not complain... (TS 2.4.1)
export const SourceFileBase: Constructor<ModuledNode> & Constructor<StatementedNode> & Constructor<TextInsertableNode> & typeof Node = ModuledNode(
    TextInsertableNode(StatementedNode(Node))
);
export class SourceFile extends SourceFileBase<ts.SourceFile> {
```
But the resulting declaration kind of does not.
You have that comment about needing to explicitly type this in TS 2.4.1. Might it be worth revisiting this?
I tried to take a look, but I do not have enough knowledge about the codebase and it quickly escalated when I tried to change it :smiley: 

In regards to IntelliJ/Webstorm:
The experience with Typescript is generally excellent.
Good performance, great auto-completion and good refactoring support.
Especially the move/extract stuff becomes a nice productivity boon once you get used to it.

**Comment by arxenix** [comment]
I'm having this same issue. Haven't had any problems with IDEA typescript integration thus far

**Comment by dsherret** [comment]
@arxenix @LeonardKoch I just tried this out in WebStore 2019.1.2 (WS-191.7141.49 -- May 7th) and it seems to work fine now. What versions are you using?

What I did find, is it takes about a second for the completion list to populate entirely with WebStorm. I also found that if I typed enough of the method name it wouldn't show up at all.

![delete](https://user-images.githubusercontent.com/1609021/58028432-b6d62680-7ae8-11e9-819a-6f87d6f1e725.gif)

So I'm going to close this as it seems to be working now. Hopefully the experience WebStorm will continue to improve and I'm glad to hear it's not as much of a pain as I found it to be 3 years ago.
